File: CSharp/OpenTraceability/Queries/EPCISTestServerClient.cs
using System;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Threading.Tasks;
using OpenTraceability.Interfaces;
using OpenTraceability.Mappers;
using OpenTraceability.Models.Events;
using OpenTraceability.Models.Identifiers;
using OpenTraceability.Utility;

namespace OpenTraceability.Queries
{
    /// <summary>
    /// A client for talking to an Open Traceability Test Server for
    /// posting and querying traceability data.
    /// </summary>
    public class EPCISTestServerClient
    {
        protected string _baseURL;
        protected EPCISDataFormat _format;
        protected EPCISVersion _version;
        protected string _apiKey;
        protected string _dataset;

        /// <summary>
        /// Creates a new client. Each client instance is bound to a single dataset on the server,
        /// identified by the X-Dataset-Id header. When no dataset id is provided a unique one is
        /// generated so that callers get an isolated dataset by default.
        /// </summary>
        public EPCISTestServerClient(string baseURL, string apiKey, EPCISDataFormat format, EPCISVersion version, string? datasetID = null)
        {
            _baseURL = baseURL;
            _version = version;
            _format = format;
            _apiKey = apiKey;
            _dataset = string.IsNullOrEmpty(datasetID) ? Guid.NewGuid().ToString() : datasetID!;
        }

        /// <summary>
        /// Posts an EPCIS Document to the Test Server under this client's dataset.
        /// </summary>
        /// <param name="doc">The EPCIS document to post.</param>
        /// <returns>The dataset id the data was posted under. Use this when querying for events after.</returns>
        public async Task<string> Post(EPCISDocument doc)
        {
            string url = $"{_baseURL.TrimEnd('/')}/epcis/events";

            IEPCISDocumentMapper mapper = OpenTraceabilityMappers.EPCISDocument.XML;
            string contentType = "application/xml";
            if (_format == EPCISDataFormat.JSON)
            {
                contentType = "application/json";
                mapper = OpenTraceabilityMappers.EPCISDocument.JSON;
            }

            using (var clientItem = HttpClientPool.GetClient())
            {
                var client = clientItem.Value;

                HttpRequestMessage request = new HttpRequestMessage();
                request.RequestUri = new Uri(url);
                request.Headers.Add("X-API-Key", _apiKey);
                request.Headers.Add("X-Dataset-Id", _dataset);

                if (_version == EPCISVersion.V1)
                {
                    request.Headers.Add("Accept", "application/xml");
                    request.Headers.Add("GS1-EPCIS-Version", "1.2");
                    request.Headers.Add("GS1-EPCIS-Min", "1.2");
                    request.Headers.Add("GS1-EPCIS-Max", "1.2");
                    request.Headers.Add("GS1-CBV-Version", "1.2");
                    request.Headers.Add("GS1-CBV-XML-Format", "ALWAYS_URN");
                }
                else if (_version == EPCISVersion.V2)
                {
                    if (_format == EPCISDataFormat.XML)
                    {
                        request.Headers.Add("Accept", "application/xml");
                    }
                    else
                    {
                        request.Headers.Add("Accept", "application/json");
                    }
                    request.Headers.Add("GS1-EPCIS-Version", "2.0");
                    request.Headers.Add("GS1-EPCIS-Min", "2.0");
                    request.Headers.Add("GS1-EPCIS-Max", "2.0");
                    request.Headers.Add("GS1-CBV-Version", "2.0");
                    request.Headers.Add("GS1-CBV-XML-Format", "ALWAYS_URN");
                }
                else
                {
                    throw new Exception($"Unrecognized EPCISVersion {_version} on the options.");
                }


                StringContent content = new StringContent(mapper.Map(doc), System.Text.Encoding.UTF8, contentType);
                request.Content = content;
                request.Method = HttpMethod.Post;

                var response = await client.SendAsync(request);
                if (!response.IsSuccessStatusCode)
                {
                    string contentStr = await response.Content.ReadAsStringAsync();
                    throw new Exception($"{(int)response.StatusCode} - {response.StatusCode} - {contentStr}");
                }

                return _dataset;
            }
        }

        /// <summary>
        /// Queries this client's dataset for events that match the parameters.
        /// </summary>
        /// <param name="parameters">The EPCIS Query parameters.</param>
        /// <returns>The EPCIS Query results.</returns>
        public async Task<EPCISQueryResults> QueryEvents(EPCISQueryParameters parameters)
        {
            using (var clientItem = HttpClientPool.GetClient())
            {
                var client = clientItem.Value;
                EPCISQueryInterfaceOptions options = BuildQueryOptions();
                return await EPCISTraceabilityResolver.QueryEvents(options, parameters, client);
            }
        }

        /// <summary>
        /// Queries and performs a traceback against this client's dataset given the EPC.
        /// </summary>
        /// <param name="epc">The EPC to perform the traceback on.</param>
        /// <returns>The epcis query results.</returns>
        public async Task<EPCISQueryResults> Traceback(EPC epc)
        {
            using (var clientItem = HttpClientPool.GetClient())
            {
                var client = clientItem.Value;
                EPCISQueryInterfaceOptions options = BuildQueryOptions();
                return await EPCISTraceabilityResolver.Traceback(options, epc, client);
            }
        }

        /// <summary>
        /// Resolves all the unknown master data in the EPCIS document.
        /// </summary>
        /// <param name="doc">The EPCIS document to resolve the master data for.</param>
        public async Task ResolveMasterData(EPCISBaseDocument doc)
        {
            using (var clientItem = HttpClientPool.GetClient())
            {
                var client = clientItem.Value;
                DigitalLinkQueryOptions options = BuildDigitalLinkOptions();
                await MasterDataResolver.ResolveMasterData(options, doc, client);
            }
        }

        /// <summary>
        /// Builds the EPCIS query interface options for this client's dataset, including the
        /// API key and X-Dataset-Id header.
        /// </summary>
        protected EPCISQueryInterfaceOptions BuildQueryOptions()
        {
            return new EPCISQueryInterfaceOptions()
            {
                URL = new Uri($"{_baseURL.TrimEnd('/')}/epcis"),
                Format = _format,
                Version = _version,
                APIKey = _apiKey,
                Headers = { ["X-Dataset-Id"] = _dataset },
                EnableStackTrace = true
            };
        }

        /// <summary>
        /// Builds the digital link query options for this client's dataset, including the
        /// API key and X-Dataset-Id header.
        /// </summary>
        protected DigitalLinkQueryOptions BuildDigitalLinkOptions()
        {
            return new DigitalLinkQueryOptions()
            {
                URL = new Uri($"{_baseURL.TrimEnd('/')}/digitallink"),
                APIKey = _apiKey,
                Headers = { ["X-Dataset-Id"] = _dataset },
                EnableStackTrace = true
            };
        }
    }
}

File: JDK11/opentraceability/src/main/java/opentraceability/queries/epcistestserverclient.java
package opentraceability.queries;

import java.net.URI;
import java.net.URL;
import java.util.UUID;
import java.util.concurrent.TimeUnit;

import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

import opentraceability.interfaces.IEPCISDocumentMapper;
import opentraceability.models.documents.EPCISBaseDocument;
import opentraceability.models.model.enums.EPCISDataFormat;
import opentraceability.models.model.enums.EPCISVersion;
import opentraceability.models.model.interfaces.EPC;
import opentraceability.models.request.*;
import opentraceability.models.results.EPCISQueryResults;
import opentraceability.mappers.OpenTraceabilityMappers;
import opentraceability.utility.HttpClientPool;


public class EPCISTestServerClient {

    private String _baseURL;
    private EPCISDataFormat _format;
    private EPCISVersion _version;

    private OkHttpClient.Builder _okHttpBuilder = new OkHttpClient.Builder()
            .connectTimeout(30, TimeUnit.SECONDS)
            .writeTimeout(30, TimeUnit.SECONDS)
            .readTimeout(30, TimeUnit.SECONDS);

    public EPCISTestServerClient(String baseURL, EPCISDataFormat format, EPCISVersion version) {
        _baseURL = baseURL;
        _format = format;
        _version = version;
    }

    private OkHttpClient getClient() {
        return _okHttpBuilder.build();
    }

    public String postEPCISDocument(EPCISDocument document) throws Exception {
        String baseUrl = _baseURL.trim().endsWith("/") ? _baseURL.substring(0, _baseURL.length()-1) : _baseURL;
        String actualBlobId = UUID.randomUUID().toString();
        String url = baseUrl + "/epcis/" + actualBlobId + "/events";

        IEPCISDocumentMapper mapper = _format == EPCISDataFormat.JSON ?
                OpenTraceabilityMappers.EPCISDocument.JSON :
                OpenTraceabilityMappers.EPCISDocument.XML;

        String contentType = _format == EPCISDataFormat.JSON ? "application/json" : "application/xml";

        OkHttpClient client = getClient();
        Request request = new Request.Builder().url(url)
            .header("Accept", _format == EPCISDataFormat.XML ? "application/xml" : "application/json")
            .header("GS1-EPCIS-Version", _version == EPCISVersion.V1 ? "1.2" : "2.0")
            .header("GS1-EPCIS-Min", _version == EPCISVersion.V1 ? "1.2" : "2.0")
            .header("GS1-EPCIS-Max", _version == EPCISVersion.V1 ? "1.2" : "2.0")
            .header("GS1-CBV-Version", _version == EPCISVersion.V1 ? "1.2" : "2.0")
            .header("GS1-CBV-XML-Format", "ALWAYS_URN")
            .post(RequestBody.create(MediaType.parse(contentType), mapper.map(document)))
            .build();

        Response response = client.newCall(request).execute();
        if (!response.isSuccessful()) {
            String contentStr = response.body().string();
            throw new Exception(String.format("%d - %s - %s",response.code(), response.message(), contentStr));
        }
        return actualBlobId;
    }


    public EPCISQueryResults queryEvents(String blobId, EPCISQueryParameters parameters) throws Exception {
        String url = _baseURL.trim() + "/epcis/" + blobId;

        EPCISQueryInterfaceOptions options = new EPCISQueryInterfaceOptions();
        options.URL = new URI(url);
        options.Format = _format;
        options.Version = _version;
        options.EnableStackTrace = true;

        return EPCISTraceabilityResolver.queryEvents(options,parameters,getClient());
    }

    public EPCISQueryResults traceback(String blobId, EPC epc) throws Exception {
        String url = _baseURL.trim() + "/epcis/" + blobId;

        EPCISQueryInterfaceOptions options = new EPCISQueryInterfaceOptions();
        options.url = new URI(url);
        options.format = _format;
        options.version = _version;
        options.enableStackTrace = true;

        return EPCISTraceabilityResolver.traceback(options, epc, getClient());
    }

    public void resolveMasterData(String blobId, EPCISBaseDocument document) throws Exception {
        String url = _baseURL.trim() + "/digitallink/" + blobId;

        DigitalLinkQueryOptions options = new DigitalLinkQueryOptions();
        options.url = new URI(url);
        options.EnableStackTrace = true;

        MasterDataResolver.resolveMasterData(options, document, getClient());
    }
}